gdnative-core/src/export/class_registry.rs

```diff
@@ -1,26 +1,50 @@
-use crate::export::NativeClass;
+use std::any::TypeId;
+use std::borrow::Cow;
+use std::collections::HashMap;
+
 use once_cell::sync::Lazy;
 use parking_lot::RwLock;
-use std::any::TypeId;
-use std::collections::HashSet;
 
-static CLASS_REGISTRY: Lazy<RwLock<HashSet<TypeId>>> = Lazy::new(|| RwLock::new(HashSet::new()));
+use crate::export::NativeClass;
+
+static CLASS_REGISTRY: Lazy<RwLock<HashMap<TypeId, ClassInfo>>> =
+    Lazy::new(|| RwLock::new(HashMap::new()));
+
+pub(crate) struct ClassInfo {
+    pub name: Cow<'static, str>,
+}
 
 /// Can be used to validate whether or not `C` has been added using `InitHandle::add_class<C>()`
 /// Returns true if added otherwise false.
 #[inline]
 pub(crate) fn is_class_registered<C: NativeClass>() -> bool {
     let type_id = TypeId::of::<C>();
-    CLASS_REGISTRY.read().contains(&type_id)
+    CLASS_REGISTRY.read().contains_key(&type_id)
+}
+
+/// Access the [`ClassInfo`] of the class `C`.
+#[inline]
+pub(crate) fn with_class_info<C: NativeClass, F, R>(f: F) -> Option<R>
+where
+    F: FnOnce(&ClassInfo) -> R,
+{
+    CLASS_REGISTRY.read().get(&TypeId::of::<C>()).map(f)
+}
+
+/// Returns the NativeScript name of the class `C` if it is registered. Returns a best-effort
+/// description of the type for error reporting otherwise.
+#[inline]
+pub(crate) fn class_name_or_default<C: NativeClass>() -> Cow<'static, str> {
+    with_class_info::<C, _, _>(|i| i.name.clone())
+        .unwrap_or_else(|| Cow::Borrowed(std::any::type_name::<C>()))
```

Comment:
`any::type_name::<C>()` might contain module prefixes and `::`, depending on the implementation.

Is this not an issue for a Godot class names, e.g. [here](https://github.com/godot-rust/godot-rust/pull/847/files#diff-dca4e8c12171c30bb0feffc7779771ea48fecad755e2e88daf40dea594eaf3afR151)? If yes, we might want to `rfind()` the last `::` and just consider the name after that.

Reply:
No, it's not an issue. I have been using `any::type_name` in generic `class_name` implementations for some time, and so far there has been no problems (except the types being hard to name outside Rust).

Reply:
Ah, good to know.

It seems like the `any::type_name()` branch is only used when the class is not registered, the comment says for error messages. Is it not used for any _logic_ in GDScript?

For diagnostics, any representation should be OK. But if the user needs to _name_ a rust-qualified type in GDScript, I see some concerns:
* Modules are a foreign concept to GDScript, all existing classes are unqualified
* It breaks with the current behavior, where classes are exported as-is
* While type hints are somewhat broken in current GDScript, this will be fixed in Godot 4, and then people likely want to use short (unqualified) identifiers
* The main reason to qualify classes would be avoiding naming conflicts, which only few users might want
* The `any::type_name()` representation is not portable/guaranteed, so it will be harder to write robust code

As mentioned, if it's for diagnostic purposes only, disregard all the above points 🙂

Reply:
It is indeed for diagnostic purposes only. I have changed the implementation of `maybe_emplace` to make that clear.

Reply:
Thanks for the clarification!

Should we maybe stress in the `class_name_or_default()` doc that the returned string should not be used for Godot class names, only as diagnostic?

It already says:
```rs
/// Returns the NativeScript name of the class `C` if it is registered. Returns a best-effort
/// description of the type for error reporting otherwise.
```
but maybe could be something along (feel free to reword):
```rs
/// Returns the NativeScript name of the class `C` if it is registered, or a best-effort description
/// of the type otherwise.
///
/// The returned string should only be used for diagnostic purposes, not for types that the user
/// has to name explicitly. The format is not guaranteed.
```

Reply:
That's fine. Updated.